[GPU Process] [FormControls] Add ControlPart for MenuListButton

shallawa opened the pull request:
#### 5a3f07e17bd7743dd41a35fda78c9cce3a724a47
<pre>
[GPU Process] [FormControls] Add ControlPart for MenuListButton
<a href="https://bugs.webkit.org/show_bug.cgi?id=249988">https://bugs.webkit.org/show_bug.cgi?id=249988</a>
rdar://103803444

Reviewed by Aditya Keerthi.

This ControlPart will handle drawing the button and background of the MenuList
if the zoomFactor != 100% (i.e. &lt;select style=&quot;zoom: 200%;&quot;&gt;...&lt;/select&gt;). For
macOS, the GraphicsContext APIs will be used to draw the platform control.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/platform/graphics/FloatRect.h:
(WebCore::FloatRect::contract):
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/MenuListButtonPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::createPlatformColorWell):
(WebCore::ControlFactoryMac::createPlatformMenuListButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm: Added.
(WebCore::MenuListButtonMac::MenuListButtonMac):
(WebCore::topGradientInterpolate):
(WebCore::bottomGradientInterpolate):
(WebCore::mainGradientInterpolate):
(WebCore::drawMenuListBackground):
(WebCore::MenuListButtonMac::draw):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::ensureControlPart):
(WebCore::RenderBox::ensureControlPartForRenderer):
(WebCore::RenderBox::ensureControlPartForBorderOnly):
(WebCore::RenderBox::ensureControlPartForDecorations):
(WebCore::RenderBox::paintBoxDecorations):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::canCreateControlPartForDecorations const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForDecorations const):
(WebCore::TopGradientInterpolate): Deleted.
(WebCore::BottomGradientInterpolate): Deleted.
(WebCore::MainGradientInterpolate): Deleted.
(WebCore::RenderThemeMac::paintMenuListButtonGradients): Deleted.
(WebCore::RenderThemeMac::paintMenuListButtonDecorations): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/258644@main">https://commits.webkit.org/258644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04ef9dbb3dda032c14f0cdd10a76752521103e88

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111856 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172078 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106557 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2623 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94865 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109563 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108368 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9757 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/37415 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24477 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5178 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/25896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5338 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2347 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11358 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45385 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5932 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7070 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
```diff
@@ -42,6 +42,13 @@ class ControlMac : public PlatformControl {
+    static constexpr float m_baseSpaceBetweenArrows = 2.0f;
+    static constexpr int m_arrowPaddingBefore = 6;
+    static constexpr int m_arrowPaddingAfter = 6;
+
```
Can these just be in `MenuListButtonMac.mm`?

shallawa replied:
They are moved to `MenuListButtonMac::draw()`

pxlcoder commented on Source/WebCore/platform/graphics/controls/ControlStyle.h:
```diff
@@ -56,7 +57,9 @@ struct ControlStyle {
     OptionSet<State> states;
     unsigned fontSize { 12 };
     float zoomFactor { 1 };
+    Color actionColor;
```
Is this specific to `MenuListButton`?

shallawa replied:
No it is also used by `SliderTrackPart::drawTicks(`). It is now called `textColor`.

pxlcoder commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1579,28 +1579,32 @@ BackgroundBleedAvoidance RenderBox::determineBackgroundBleedAvoidance(GraphicsCo
-    if (!rareData.controlPart)
+    auto type = style().effectiveAppearance();
+
+    // Some form-controls may change because of zoominh without recreating
```
Typo: `zooming`.

shallawa committed:
[GPU Process] [FormControls] Add ControlPart for MenuListButton
https://bugs.webkit.org/show_bug.cgi?id=249988
rdar://103803444

Reviewed by Aditya Keerthi.

This ControlPart will handle drawing the button and background of the MenuList
if the zoomFactor != 100% (i.e. <select style="zoom: 200%;">...</select>). For
macOS, the GraphicsContext APIs will be used to draw the platform control.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/platform/graphics/FloatRect.h:
(WebCore::FloatRect::contract):
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/MenuListButtonPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::createPlatformColorWell):
(WebCore::ControlFactoryMac::createPlatformMenuListButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm: Added.
(WebCore::MenuListButtonMac::MenuListButtonMac):
(WebCore::topGradientInterpolate):
(WebCore::bottomGradientInterpolate):
(WebCore::mainGradientInterpolate):
(WebCore::drawMenuListBackground):
(WebCore::MenuListButtonMac::draw):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::ensureControlPart):
(WebCore::RenderBox::ensureControlPartForRenderer):
(WebCore::RenderBox::ensureControlPartForBorderOnly):
(WebCore::RenderBox::ensureControlPartForDecorations):
(WebCore::RenderBox::paintBoxDecorations):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::canCreateControlPartForDecorations const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForDecorations const):
(WebCore::TopGradientInterpolate): Deleted.
(WebCore::BottomGradientInterpolate): Deleted.
(WebCore::MainGradientInterpolate): Deleted.
(WebCore::RenderThemeMac::paintMenuListButtonGradients): Deleted.
(WebCore::RenderThemeMac::paintMenuListButtonDecorations): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/258644@main
